### PR TITLE
Add check for TypeBounds on argument substitutions

### PR DESCRIPTION
**Pull request by philberty**
TypeParameters contain specified bounds, when we substitute them we need
to ensure that the argument satisfies those bounds otherwise... it fails
the contract.

Addreses #440

### CODE REVIEW

File: gcc/rust/typecheck/rust-tyty.cc [hunk 1]
```diff
@@ -27,13 +27,41 @@
...
+
+  std::vector<std::reference_wrapper<Resolver::TraitReference>> probed
+    = Resolver::TypeBoundsProbe::Probe (this);
+  for (const Resolver::TraitReference &bound : probed)
+    {
+      bool found = bound.get_mappings ().get_defid ()
+		   == query->get_mappings ().get_defid ();
+      if (found)
+	return true;
+    }
```
Comment (CohenArthur) [inline gcc/rust/typecheck/rust-tyty.cc:60]: I'm not sure I'm understanding this code correctly. Does this mean that as soon as you find a matching bound, the function returns true? In that case, is it intended (for now) that multiple bounds are not verified correctly?

For example, the following example compiles succesfully:

```rust
trait Foo {
    fn default() -> i32;
}

trait Bar {
    fn not_default() -> i32;
}

struct Test(i32);

impl Foo for Test {
    fn default() -> i32 {
        1234
    }
}

fn type_bound_test<T: Foo + Bar>() -> i32 { // <T: Bar + Foo> also compiles
    T::default()
}

fn main() {
    let a = type_bound_test::<Test>();
}
```

Comment (philberty) [reply gcc/rust/typecheck/rust-tyty.cc:60]: Let me add that test case to verify this.

But the code is split up into 2 parts this function is about checking each bound the driver of this is actually here: 

https://github.com/Rust-GCC/gccrs/pull/611/files#diff-03fbb543dd5e74fc2332fa2e5c9c9a91ed6d46d86e7bf0310a8796fae439d26cR256-R275

So we iterate the bounds of the TypePaameter then verify that each of these are satisfied by the substitution.

Comment (CohenArthur) [reply gcc/rust/typecheck/rust-tyty.cc]: Ah, right thanks. I see how the code works exactly now. Glad that multiple bounds are working :D Good job!

**Comment by philberty** [comment]
bors cancel

**Comment by philberty** [comment]
bors merge

**Pull request merged**